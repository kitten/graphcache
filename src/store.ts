import invariant from 'invariant';
import { DocumentNode } from 'graphql';
import * as Pessimism from 'pessimism';

import {
  EntityField,
  Link,
  ResolverConfig,
  DataField,
  Variables,
  Data,
  UpdatesConfig,
  OptimisticMutationConfig,
  KeyingConfig,
} from './types';

import { keyOfEntity, joinKeys, keyOfField } from './helpers';
import { startQuery } from './operations/query';
import { writeFragment, startWrite } from './operations/write';
import { invalidate } from './operations/invalidate';

interface Ref<T> {
  current: null | T;
}

const currentDependencies: Ref<Set<string>> = { current: null };
const currentOptimisticKey: Ref<number> = { current: null };

// Resolve a ref value or throw when we're outside of a store run
const refValue = <T>(ref: Ref<T>): T => {
  invariant(
    ref.current !== null,
    'Invalid Cache call: The cache may only be accessed or mutated during' +
      'operations like write or query, or as part of its resolvers, updaters, ' +
      'or optimistic configs.'
  );

  return ref.current as T;
};

// Initialise a store run by resetting its internal state
export const initStoreState = (optimisticKey: null | number) => {
  currentDependencies.current = new Set();
  currentOptimisticKey.current = optimisticKey;
};

// Finalise a store run by clearing its internal state
export const clearStoreState = () => {
  currentDependencies.current = null;
  currentOptimisticKey.current = null;
};

export const getCurrentDependencies = () => refValue(currentDependencies);

// Add a dependency to the internal store state
export const addDependency = (dependency: string) => {
  (currentDependencies.current as Set<string>).add(dependency);
};

const mapSet = <T>(map: Pessimism.Map<T>, key: string, value: T) => {
  const optimisticKey = currentOptimisticKey.current || 0;
  return Pessimism.setOptimistic(map, key, value, optimisticKey);
};

// Used to remove a value from a Map optimistially (possible by setting it to undefined)
const mapRemove = <T>(map: Pessimism.Map<T>, key: string) => {
  const optimisticKey = currentOptimisticKey.current || 0;
  return optimisticKey
    ? Pessimism.setOptimistic(map, key, undefined, optimisticKey)
    : Pessimism.remove(map, key);
};

export class Store {
  records: Pessimism.Map<EntityField>;
  links: Pessimism.Map<Link>;

  resolvers: ResolverConfig;
  updates: UpdatesConfig;
  optimisticMutations: OptimisticMutationConfig;
  keys: KeyingConfig;

  constructor(
    resolvers?: ResolverConfig,
    updates?: Partial<UpdatesConfig>,
    optimisticMutations?: OptimisticMutationConfig,
    keys?: KeyingConfig
  ) {
    this.records = Pessimism.asMutable(Pessimism.make());
    this.links = Pessimism.asMutable(Pessimism.make());
    this.resolvers = resolvers || {};
    this.updates = {
      Mutation: (updates && updates.Mutation) || {},
      Subscription: (updates && updates.Subscription) || {},
    } as UpdatesConfig;
    this.optimisticMutations = optimisticMutations || {};
    this.keys = keys || {};
  }

  keyOfEntity(data: Data) {
    const { __typename: typename } = data;
    if (typename !== undefined && this.keys[typename] !== undefined) {
      return this.keys[typename](data);
    } else {
      return keyOfEntity(data);
    }
  }

  clearOptimistic(optimisticKey: number) {
    this.records = Pessimism.clearOptimistic(this.records, optimisticKey);
    this.links = Pessimism.clearOptimistic(this.links, optimisticKey);
  }

  getRecord(fieldKey: string): EntityField {
    return Pessimism.get(this.records, fieldKey);
  }

  removeRecord(fieldKey: string) {
    return (this.records = mapRemove(this.records, fieldKey));
  }

  writeRecord(field: EntityField, fieldKey: string) {
    return (this.records = mapSet(this.records, fieldKey, field));
  }

  getField(
    entityKey: string,
    fieldName: string,
    args?: Variables
  ): EntityField {
    const fieldKey = joinKeys(entityKey, keyOfField(fieldName, args));
    return this.getRecord(fieldKey);
  }

  writeField(
    field: EntityField,
    entityKey: string,
    fieldName: string,
    args?: Variables
  ) {
    const fieldKey = joinKeys(entityKey, keyOfField(fieldName, args));
    return (this.records = mapSet(this.records, fieldKey, field));
  }

  getLink(key: string): undefined | Link {
    return Pessimism.get(this.links, key);
  }

  removeLink(key: string) {
    return (this.links = mapRemove(this.links, key));
  }

  writeLink(link: Link, key: string) {
    return (this.links = mapSet(this.links, key, link));
  }

  resolveValueOrLink(fieldKey: string): DataField {
    const fieldValue = this.getRecord(fieldKey);
    // Undefined implies a link OR incomplete data.
    // A value will imply that we are just fetching a field like date.
    if (fieldValue !== undefined) return fieldValue;

    // This can be an array OR a string OR undefined again
    const link = this.getLink(fieldKey);
    return link ? link : null;
  }

  resolve(entity: Data | string, field: string, args?: Variables): DataField {
    if (typeof entity === 'string') {
      addDependency(entity);
      return this.resolveValueOrLink(joinKeys(entity, keyOfField(field, args)));
    } else {
      // This gives us __typename:key
      const entityKey = this.keyOfEntity(entity);
      if (entityKey === null) return null;
      addDependency(entityKey);
      return this.resolveValueOrLink(
        joinKeys(entityKey, keyOfField(field, args))
      );
    }
  }

  invalidateQuery(dataQuery: DocumentNode, variables: Variables) {
    invalidate(this, { query: dataQuery, variables });
  }

  hasField(key: string): boolean {
    return this.getRecord(key) !== undefined || this.getLink(key) !== undefined;
  }

  updateQuery(
    dataQuery: DocumentNode,
    updater: (data: Data | null) => null | Data
  ): void {
    const { data, completeness } = startQuery(this, { query: dataQuery });
    const input = completeness === 'EMPTY' ? null : data;
    const output = updater(input);
    if (output !== null) {
      startWrite(this, { query: dataQuery }, output);
    }
  }

  writeFragment(dataFragment: DocumentNode, data: Data): void {
    writeFragment(this, dataFragment, data);
  }
}
