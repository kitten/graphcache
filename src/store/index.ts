export {
  makeDict,
  initDataState,
  clearDataState,
  clearOptimistic,
  getCurrentDependencies,
} from './data';

export * from './keys';
export * from './store';
