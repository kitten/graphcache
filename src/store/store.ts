import { DocumentNode } from 'graphql';
import { createRequest } from 'urql/core';

import {
  Cache,
  FieldInfo,
  ResolverConfig,
  DataField,
  Variables,
  Data,
  QueryInput,
  UpdatesConfig,
  OptimisticMutationConfig,
  KeyingConfig,
} from '../types';

import { read, readFragment } from '../operations/query';
import { writeFragment, startWrite } from '../operations/write';
import { invalidate } from '../operations/invalidate';
import { SchemaPredicates } from '../ast';
import { keyOfField } from './keys';
import * as InMemoryData from './data';

type RootField = 'query' | 'mutation' | 'subscription';

export class Store implements Cache {
  data: InMemoryData.InMemoryData;

  resolvers: ResolverConfig;
  updates: UpdatesConfig;
  optimisticMutations: OptimisticMutationConfig;
  keys: KeyingConfig;
  schemaPredicates?: SchemaPredicates;

  rootFields: { query: string; mutation: string; subscription: string };
  rootNames: { [name: string]: RootField };

  constructor(
    schemaPredicates?: SchemaPredicates,
    resolvers?: ResolverConfig,
    updates?: Partial<UpdatesConfig>,
    optimisticMutations?: OptimisticMutationConfig,
    keys?: KeyingConfig
  ) {
    this.resolvers = resolvers || {};
    this.optimisticMutations = optimisticMutations || {};
    this.keys = keys || {};
    this.schemaPredicates = schemaPredicates;

    this.updates = {
      Mutation: (updates && updates.Mutation) || {},
      Subscription: (updates && updates.Subscription) || {},
    } as UpdatesConfig;

    if (schemaPredicates) {
      const { schema } = schemaPredicates;
      const queryType = schema.getQueryType();
      const mutationType = schema.getMutationType();
      const subscriptionType = schema.getSubscriptionType();

      const queryName = queryType ? queryType.name : 'Query';
      const mutationName = mutationType ? mutationType.name : 'Mutation';
      const subscriptionName = subscriptionType
        ? subscriptionType.name
        : 'Subscription';

      this.rootFields = {
        query: queryName,
        mutation: mutationName,
        subscription: subscriptionName,
      };

      this.rootNames = {
        [queryName]: 'query',
        [mutationName]: 'mutation',
        [subscriptionName]: 'subscription',
      };
    } else {
      this.rootFields = {
        query: 'Query',
        mutation: 'Mutation',
        subscription: 'Subscription',
      };

      this.rootNames = {
        Query: 'query',
        Mutation: 'mutation',
        Subscription: 'subscription',
      };
    }

    this.data = InMemoryData.make(this.getRootKey('query'));
  }

  gcScheduled = false;
  gc = () => {
    InMemoryData.gc(this.data);
    this.gcScheduled = false;
  };

  keyOfField = keyOfField;

  getRootKey(name: RootField) {
    return this.rootFields[name];
  }

  keyOfEntity(data: Data) {
    const { __typename: typename, id, _id } = data;
    if (!typename) {
      return null;
    } else if (this.rootNames[typename] !== undefined) {
      return typename;
    }

    let key: string | null | void;
    if (this.keys[typename]) {
      key = this.keys[typename](data);
    } else if (id !== undefined && id !== null) {
      key = `${id}`;
    } else if (_id !== undefined && _id !== null) {
      key = `${_id}`;
    }

    return key ? `${typename}:${key}` : null;
  }

  resolveFieldByKey(entity: Data | string | null, fieldKey: string): DataField {
    const entityKey =
      entity !== null && typeof entity !== 'string'
        ? this.keyOfEntity(entity)
        : entity;
    if (entityKey === null) return null;
    const fieldValue = InMemoryData.readRecord(entityKey, fieldKey);
    if (fieldValue !== undefined) return fieldValue;
    const link = InMemoryData.readLink(entityKey, fieldKey);
    return link ? link : null;
  }

  resolve(
    entity: Data | string | null,
    field: string,
    args?: Variables
  ): DataField {
    return this.resolveFieldByKey(entity, keyOfField(field, args));
  }

  invalidateQuery(query: string | DocumentNode, variables?: Variables) {
    invalidate(this, createRequest(query, variables));
  }

  inspectFields(entity: Data | string | null): FieldInfo[] {
    const entityKey =
      entity !== null && typeof entity !== 'string'
        ? this.keyOfEntity(entity)
        : entity;
    return entityKey !== null ? InMemoryData.inspectFields(entityKey) : [];
  }

  updateQuery(
    input: QueryInput,
    updater: (data: Data | null) => Data | null
  ): void {
    const request = createRequest(input.query, input.variables);
    const output = updater(this.readQuery(request as QueryInput));
    if (output !== null) {
      startWrite(this, request, output);
    }
  }

  readQuery(input: QueryInput): Data | null {
    return read(this, createRequest(input.query, input.variables)).data;
  }

  readFragment(
    dataFragment: DocumentNode,
    entity: string | Data,
    variables?: Variables
  ): Data | null {
    return readFragment(this, dataFragment, entity, variables);
  }

  writeFragment(
    dataFragment: DocumentNode,
    data: Data,
    variables?: Variables
  ): void {
    writeFragment(this, dataFragment, data, variables);
  }
}
