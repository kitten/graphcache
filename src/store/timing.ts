export const defer: (fn: () => void) => void =
  process.env.NODE_ENV === 'production' && typeof Promise !== 'undefined'
    ? Promise.prototype.then.bind(Promise.resolve())
    : fn => setTimeout(fn, 0);
