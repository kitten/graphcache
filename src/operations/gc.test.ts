import gql from 'graphql-tag';
import { Store } from '../store';
import { gc, write } from './index';

const gqlQuery = gql`
  {
    __typename
    todos {
      __typename
      id
      text
      complete
    }
    todo {
      __typename
      id
      text
      complete
      author {
        __typename
        name
      }
    }
  }
`;

it('cleans up unreachable records and links', () => {
  const store = new Store();

  write(
    store,
    { query: gqlQuery },
    {
      __typename: 'Query',
      todos: [
        {
          id: '0',
          text: 'Go to the shops',
          complete: false,
          __typename: 'Todo',
        },
        {
          id: '1',
          text: 'Pick up the kids',
          complete: true,
          __typename: 'Todo',
        },
        { id: '2', text: 'Install urql', complete: false, __typename: 'Todo' },
      ],
      todo: {
        __typename: 'Todo',
        id: '2',
        text: 'Install urql',
        complete: false,
        author: {
          __typename: 'Author',
          name: 'Thomas',
        },
      },
    }
  );

  let json = store.serialize();

  expect(json.links['Query.todos']).toEqual(['Todo:0', 'Todo:1', 'Todo:2']);
  expect(json.records['Todo:1']).toMatchObject({ __typename: 'Todo' });
  expect(json.records['Todo:2.author']).toMatchObject({ __typename: 'Author' });

  write(
    store,
    { query: gqlQuery },
    {
      __typename: 'Query',
      todos: null,
      todo: {
        __typename: 'Todo',
        id: '2',
        text: 'Install urql',
        complete: false,
        author: {
          __typename: 'Author',
          name: 'Thomas',
        },
      },
    }
  );

  json = store.serialize();

  // Same as above; everything is still reachable but `query.todos` is gone
  expect(json.records['Todo:1']).toMatchObject({ __typename: 'Todo' });
  expect(json.records['Todo:2.author']).toMatchObject({ __typename: 'Author' });

  gc(store);
  json = store.serialize();

  expect(json).toMatchSnapshot();
  expect(json.records['Todo:1']).toBe(undefined);
  expect(json.records['Todo:2']).not.toBe(undefined);
  expect(json.records['Todo:2.author']).toMatchObject({ __typename: 'Author' });

  // Remove all results
  write(
    store,
    { query: gqlQuery },
    {
      __typename: 'Query',
      todos: null,
      todo: null,
    }
  );

  gc(store);
  json = store.serialize();

  expect(json).toEqual({
    records: {
      Query: {
        __typename: 'Query',
        todos: undefined,
        todo: undefined,
      },
    },
    links: {
      'Query.todo': null,
      'Query.todos': null,
    },
  });
});
