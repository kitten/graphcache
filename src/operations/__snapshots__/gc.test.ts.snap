// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cleans up unreachable records and links 1`] = `
Object {
  "links": Object {
    "Query.todo": "Todo:2",
    "Query.todos": null,
    "Todo:2.author": "Todo:2.author",
  },
  "records": Object {
    "Query": Object {
      "__typename": "Query",
      "todo": undefined,
      "todos": undefined,
    },
    "Todo:2": Object {
      "__typename": "Todo",
      "author": undefined,
      "complete": false,
      "id": "2",
      "text": "Install urql",
    },
    "Todo:2.author": Object {
      "__typename": "Author",
      "name": "Thomas",
    },
  },
}
`;
